### PR TITLE
transfer-sh: init at 1.6.0

### PR DESCRIPTION
**Pull request by pinpox**
###### Description of changes

Added a package and module for transfer.sh, addressing #239627 
I'm sure there are things that can be done in a cleaner way, please let me know and I'll iterate over the code until it fits the quality standards.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/misc/transfer-sh.nix [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+            TLS_LISTENER = "${cfg.address}:${builtins.toString cfg.TLS_LISTENER}";
+            TLS_LISTENER_ONLY = cfg.TLS_LISTENER_ONLY;
+            TLS_CERT_FILE = cfg.TLS_CERT_FILE;
+            TLS_PRIVATE_KEY = cfg.TLS_PRIVATE_KEY;
+            FORCE_HTTPS = cfg.FORCE_HTTPS;
+          }
+          else { });
+
+      };
+      networking.firewall.allowedTCPPorts = [ cfg.LISTENER cfg.PROFILE_LISTENER ]
```
Comment (tilpner) [inline nixos/modules/services/misc/transfer-sh.nix:411]: Altering the firewall should be opt-in, not default behaviour.

File: nixos/modules/services/misc/transfer-sh.nix [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+
+          # TLS specific options
+          (if cfg.TLS_LISTENER != null then {
+            TLS_LISTENER = "${cfg.address}:${builtins.toString cfg.TLS_LISTENER}";
+            TLS_LISTENER_ONLY = cfg.TLS_LISTENER_ONLY;
+            TLS_CERT_FILE = cfg.TLS_CERT_FILE;
+            TLS_PRIVATE_KEY = cfg.TLS_PRIVATE_KEY;
+            FORCE_HTTPS = cfg.FORCE_HTTPS;
+          }
+          else { });
```
Comment (tilpner) [inline nixos/modules/services/misc/transfer-sh.nix:408]: This `if ... then { ... } else {}` could use `lib.optionalAttrs`.

File: nixos/modules/services/misc/transfer-sh.nix [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+        serviceConfig = {
+          User = cfg.user;
+          Group = cfg.group;
+          ExecStart = "${cfg.package}/bin/transfer.sh --provider=${cfg.provider}";
+          EnvironmentFile = lib.mkIf (cfg.environmentFile != null) cfg.environmentFile;
+        };
+
+        environment =
+          let
+            boolString = s: if s then "true" else "false";
```
Comment (tilpner) [inline nixos/modules/services/misc/transfer-sh.nix:347]: This looks like a reimplementation of `lib.boolToString`.

File: pkgs/servers/transfer-sh/default.nix [hunk 4]
```diff
@@ -0,0 +1,24 @@
+{ lib, fetchFromGitHub, buildGoModule }:
+
+buildGoModule rec {
+  pname = "transfer-sh";
+  version = "1.6.0";
+
+  src = fetchFromGitHub {
+    rev = "v${version}";
+    owner = "dutchcoders";
+    repo = "transfer.sh";
```
Comment (SuperSandro2000) [inline pkgs/servers/transfer-sh/default.nix:10]:
```suggestion
    owner = "dutchcoders";
    repo = "transfer.sh";
    rev = "v${version}";
```

File: pkgs/servers/transfer-sh/default.nix [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  };
+
+  vendorSha256 = "sha256-CpEdWhvvQhvTWra7xQnXgJWKDnaod6NHkGiz4hoTy4g=";
+
+  meta = with lib; {
+    description = "Easy and fast file sharing and pastebin server with access from the command-line";
+    homepage = "https://github.com/dutchcoders/transfer.sh";
+    changelog = "https://github.com/dutchcoders/transfer.sh/releases";
+    license = licenses.mit;
+    platforms = platforms.linux;
```
Comment (SuperSandro2000) [inline pkgs/servers/transfer-sh/default.nix:21]: Why exclude darwin?

```suggestion
```

File: nixos/modules/services/misc/transfer-sh.nix [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+
+with lib;
+
+let
+  cfg = config.services.transfer-sh;
+in
+{
+  options = {
+    services.transfer-sh = {
+
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/transfer-sh.nix:11]:
```suggestion
```

File: nixos/modules/services/misc/transfer-sh.nix [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+      enable = mkEnableOption "transfer-sh setup";
+
+      package = mkOption {
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/transfer-sh.nix:14]: Please use mkPackageOption

File: nixos/modules/services/misc/transfer-sh.nix [hunk 8]
```diff
@@ -0,0 +1,434 @@
...
+            TLS_LISTENER_ONLY = cfg.TLS_LISTENER_ONLY;
+            TLS_CERT_FILE = cfg.TLS_CERT_FILE;
+            TLS_PRIVATE_KEY = cfg.TLS_PRIVATE_KEY;
+            FORCE_HTTPS = cfg.FORCE_HTTPS;
+          }
+          else { });
+
+      };
+      networking.firewall.allowedTCPPorts = [ cfg.LISTENER cfg.PROFILE_LISTENER ]
+        ++ (if cfg.TLS_LISTENER != null then [ cfg.TLS_LISTENER ] else [ ]);
```
Comment (tilpner) [inline nixos/modules/services/misc/transfer-sh.nix:412]: shorter with `lib.optionals`

File: nixos/modules/services/misc/transfer-sh.nix [hunk 9]
```diff
@@ -0,0 +1,434 @@
...
+        type = types.enum [ "s3" "storj" "gdrive" "local" ];
+      };
+
+      address = mkOption {
+        description = "address to listen on";
+        default = "127.0.0.1";
+        type = types.string;
+      };
+
+      LISTENER = mkOption {
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/transfer-sh.nix:51]: LISTENER is not a great name for the port option

File: nixos/modules/services/misc/transfer-sh.nix [hunk 10]
```diff
@@ -0,0 +1,434 @@
...
+        description = "port to use for http";
+        default = 6080;
+        type = types.nullOr types.int;
+      };
+
+      PROFILE_LISTENER = mkOption {
+        description = "port to use for profiler";
+        default = 6060;
+        type = types.nullOr types.int;
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/transfer-sh.nix:60]:
```suggestion
        type = types.port;
      };

      PROFILE_LISTENER = mkOption {
        description = "port to use for profiler";
        default = 6060;
        type = types.nullOr types.port;
```
I assume we always need at least the http port

**Comment by mweinelt** [comment]
The commit messages do not fit CONTRIBUTING.md.

File: nixos/modules/services/misc/transfer-sh.nix [hunk 11]
```diff
@@ -0,0 +1,434 @@
...
+      };
+
+      FORCE_HTTPS = mkOption {
+        description = "redirect to https";
+        default = "false";
+        type = types.nullOr types.str;
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/transfer-sh.nix:66]: Why is this not a bool?

File: nixos/modules/services/misc/transfer-sh.nix [hunk 12]
```diff
@@ -0,0 +1,434 @@
...
+      };
+
+      TLS_LISTENER = mkOption {
+        description = "port to use for https";
+        default = null;
+        type = types.nullOr types.int;
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/transfer-sh.nix:72]:
```suggestion
        type = types.nullOr types.port;
```

File: nixos/modules/services/misc/transfer-sh.nix [hunk 13]
```diff
@@ -0,0 +1,434 @@
...
+      };
+
+      TLS_LISTENER_ONLY = mkOption {
+        description = "flag to enable tls listener only";
+        default = "false";
+        type = types.nullOr types.str;
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/transfer-sh.nix:78]: also sounds like bool

File: nixos/modules/services/misc/transfer-sh.nix [hunk 14]
```diff
@@ -0,0 +1,434 @@
...
+      };
+
+      TLS_CERT_FILE = mkOption {
+        description = "path to tls certificate";
+        default = null;
+        type = types.nullOr types.str;
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/transfer-sh.nix:84]: Shouldn't this be path?

File: nixos/modules/services/misc/transfer-sh.nix [hunk 15]
```diff
@@ -0,0 +1,434 @@
...
+      };
+
+      TLS_PRIVATE_KEY = mkOption {
+        description = "path to tls private key";
+        default = null;
+        type = types.nullOr types.str;
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/transfer-sh.nix:90]: same

File: nixos/modules/services/misc/transfer-sh.nix [hunk 16]
```diff
@@ -0,0 +1,434 @@
...
+      HTTP_AUTH_PASS = mkOption {
+        description = "pass for basic http auth on upload";
+        default = null;
+        type = types.nullOr types.str;
+      };
+
+      HTTP_AUTH_HTPASSWD = mkOption {
+        description = "htpasswd file path for basic http auth on upload";
+        default = null;
+        type = types.nullOr types.str;
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/transfer-sh.nix:108]: this also sounds like path

File: nixos/modules/services/misc/transfer-sh.nix [hunk 17]
```diff
@@ -0,0 +1,434 @@
...
+          ExecStart = "${cfg.package}/bin/transfer.sh --provider=${cfg.provider}";
+          EnvironmentFile = lib.mkIf (cfg.environmentFile != null) cfg.environmentFile;
+        };
+
+        environment =
+          let
+            boolString = s: if s then "true" else "false";
+          in
+          {
+            LISTENER = "${cfg.address}:${builtins.toString cfg.LISTENER}";
```
Comment (tilpner) [inline nixos/modules/services/misc/transfer-sh.nix:350]: You can call `toString` without `builtins.`

File: nixos/modules/services/misc/transfer-sh.nix [hunk 18]
```diff
@@ -0,0 +1,434 @@
...
+      HTTP_AUTH_HTPASSWD = mkOption {
+        description = "htpasswd file path for basic http auth on upload";
+        default = null;
+        type = types.nullOr types.str;
+      };
+
+      HTTP_AUTH_IP_WHITELIST = mkOption {
+        description = "comma separated list of ips allowed to upload without being challenged an http auth";
+        default = null;
+        type = types.nullOr types.str;
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/transfer-sh.nix:114]: This should probably be a list of strings

File: nixos/modules/services/misc/transfer-sh.nix [hunk 19]
```diff
@@ -0,0 +1,434 @@
...
+      IP_WHITELIST = mkOption {
+        description = "comma separated list of ips allowed to connect to the service";
+        default = null;
+        type = types.nullOr types.str;
+      };
+
+      IP_BLACKLIST = mkOption {
+        description = "comma separated list of ips not allowed to connect to the service";
+        default = null;
+        type = types.nullOr types.str;
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/transfer-sh.nix:126]: same with those

File: nixos/modules/services/misc/transfer-sh.nix [hunk 20]
```diff
@@ -0,0 +1,434 @@
...
+      };
+
+      TEMP_PATH = mkOption {
+        description = "path to temp folder";
+        default = null;
+        type = types.nullOr types.str;
+      };
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/transfer-sh.nix:133]: This should just be hardcoded in the systemd unit

File: nixos/modules/services/misc/transfer-sh.nix [hunk 21]
```diff
@@ -0,0 +1,434 @@
...
+      PROXY_PATH = mkOption {
+        description = "path prefix when service is run behind a proxy";
+        default = null;
+        type = types.nullOr types.str;
+      };
+
+      PROXY_PORT = mkOption {
+        description = "port of the proxy when the service is run behind a proxy";
+        default = null;
+        type = types.nullOr types.int;
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/transfer-sh.nix:150]:
```suggestion
        type = types.nullOr types.port;
```

File: nixos/modules/services/misc/transfer-sh.nix [hunk 22]
```diff
@@ -0,0 +1,434 @@
...
+      EMAIL_CONTACT = mkOption {
+        description = "email contact for the front end";
+        default = null;
+        type = types.nullOr types.str;
+      };
+
+      GA_KEY = mkOption {
+        description = "google analytics key for the front end";
+        default = null;
+        type = types.nullOr types.str;
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/transfer-sh.nix:162]: I hope no one sets this one

**Review by tilpner — CHANGES_REQUESTED** [review]
The firewall bit is important, the rest are just style nitpicks

File: nixos/modules/services/misc/transfer-sh.nix [hunk 23]
```diff
@@ -0,0 +1,434 @@
...
+        description = "aws access key";
+        default = null;
+        type = types.nullOr types.str;
+      };
+
+      AWS_SECRET_KEY = mkOption {
+        description = "aws access key";
+        default = null;
+        type = types.nullOr types.str;
+      };
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/transfer-sh.nix:181]: This promotes setting configuration in an insecure way

File: nixos/modules/services/misc/transfer-sh.nix [hunk 24]
```diff
@@ -0,0 +1,434 @@
...
+        "d ${cfg.BASEDIR} 0750 ${cfg.user} ${cfg.group} - -"
+      ];
+
+      systemd.services.transfer-sh = {
+        wantedBy = [ "multi-user.target" ];
+        after = [ "network.target" ];
+        serviceConfig = {
+          User = cfg.user;
+          Group = cfg.group;
+          ExecStart = "${cfg.package}/bin/transfer.sh --provider=${cfg.provider}";
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/transfer-sh.nix:341]:
```suggestion
          ExecStart = "${lib.getExe fg.package} --provider=${cfg.provider}";
```

File: nixos/modules/services/misc/transfer-sh.nix [hunk 25]
```diff
@@ -0,0 +1,434 @@
...
+          } else { }) // (if cfg.provider == "storj" then {
+            STORJ_ACCESS = cfg.STORJ_ACCESS;
+            STORJ_BUCKET = cfg.STORJ_BUCKET;
+
+            # Options specific to google drive backend
+          } else { }) // (if cfg.provider == "gdrive" then {
+            GDRIVE_CLIENT_JSON_FILEPATH = cfg.GDRIVE_CLIENT_JSON_FILEPATH;
+            GDRIVE_LOCAL_CONFIG_PATH = cfg.GDRIVE_LOCAL_CONFIG_PATH;
+            GDRIVE_CHUNK_SIZE = cfg.GDRIVE_CHUNK_SIZE;
+          } else { }) //
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/transfer-sh.nix]: This should use lib.optionalAttrs

File: nixos/modules/services/misc/transfer-sh.nix (hunk 2 above)
Comment (SuperSandro2000) [inline nixos/modules/services/misc/transfer-sh.nix]: same here

File: nixos/modules/services/misc/transfer-sh.nix (hunk 8 above)
Comment (SuperSandro2000) [inline nixos/modules/services/misc/transfer-sh.nix]:
```suggestion
        ++ lib.optional (cfg.TLS_LISTENER != null) cfg.TLS_LISTENER;
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
Please follow the [contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#submitting-changes) when naming your commits.

Could this module use freeform settings instead?

File: nixos/modules/services/misc/transfer-sh.nix [hunk 26]
```diff
@@ -0,0 +1,422 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
```
Comment (AndersonTorres) [inline nixos/modules/services/misc/transfer-sh.nix:3]:
```suggestion
```

File: nixos/modules/services/misc/transfer-sh.nix [hunk 27]
```diff
@@ -0,0 +1,422 @@
...
+        type = types.string;
+      };
+
+      LISTENER = mkOption {
+        description = "port to use for http";
+        default = 6080;
+        type = types.nullOr types.int;
+      };
+
+      PROFILE_LISTENER = mkOption {
```
Comment (AndersonTorres) [inline nixos/modules/services/misc/transfer-sh.nix:57]: Don't use uppercase.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: nixos/modules/services/misc/transfer-sh.nix (hunk 2 above)
Comment (pinpox) [reply nixos/modules/services/misc/transfer-sh.nix]: fixed

File: nixos/modules/services/misc/transfer-sh.nix (hunk 8 above)
Comment (pinpox) [reply nixos/modules/services/misc/transfer-sh.nix]: fixed.

File: nixos/modules/services/misc/transfer-sh.nix (hunk 17 above)
Comment (pinpox) [reply nixos/modules/services/misc/transfer-sh.nix]: deleted

File: nixos/modules/services/misc/transfer-sh.nix (hunk 9 above)
Comment (pinpox) [reply nixos/modules/services/misc/transfer-sh.nix:56]: That is true, but I sticked to the upstream namings to make it easy to look up the docs. Unless you have a strong opinion against it, I'd just stick with the name to be consistent with upstream docs

File: nixos/modules/services/misc/transfer-sh.nix (hunk 23 above)
Comment (pinpox) [reply nixos/modules/services/misc/transfer-sh.nix:186]: Should I take it out? I added a warning further down

File: nixos/modules/services/misc/transfer-sh.nix (hunk 25 above)
Comment (pinpox) [reply nixos/modules/services/misc/transfer-sh.nix]: changed.

File: nixos/modules/services/misc/transfer-sh.nix (hunk 2 above)
Comment (pinpox) [reply nixos/modules/services/misc/transfer-sh.nix]: changed.

File: nixos/modules/services/misc/transfer-sh.nix (hunk 27 above)
Comment (pinpox) [reply nixos/modules/services/misc/transfer-sh.nix:61]: The settings is called like this upstream. I'd vouch for keeping it as is so the docs are consistent.

File: nixos/modules/services/misc/transfer-sh.nix (hunk 26 above)
Comment (pinpox) [reply nixos/modules/services/misc/transfer-sh.nix]: How is this not needed? If I stake it out, I'll have to add `lib.` to every mkOption, don't I?

File: nixos/modules/services/misc/transfer-sh.nix (hunk 20 above)
Comment (pinpox) [reply nixos/modules/services/misc/transfer-sh.nix:133]: Not sure how to do this. Isn't this user-configurable, how can I set it in the unit?

File: nixos/modules/services/misc/transfer-sh.nix (hunk 22 above)
Comment (pinpox) [reply nixos/modules/services/misc/transfer-sh.nix:162]: Should I take it out?

**Comment by AndersonTorres** [comment]
> Should module and pkg be separate commits or can I "init at xxx" both at once?

1. One commit for the Nixpkgs package itself
2. One commit for the NixOS module with title `nixos/<moduleName>: init` (when applicable)
3. One commit for the NixOS tests `nixos/tests/<moduleName>: init` (when applicable)

**Comment by pinpox** [comment]
> The commit messages do not fit CONTRIBUTING.md.

Should match the format now.

File: nixos/modules/services/misc/transfer-sh.nix (hunk 7 above)
Comment (pinpox) [reply nixos/modules/services/misc/transfer-sh.nix]: fixed

**Comment by Lassulus** [comment]
can we maybe put all the CAPSLOCK option inside a freeform submodule? and just document the couple ones which are important?

File: nixos/modules/services/misc/transfer-sh.nix (hunk 26 above)
Comment (nu-nu-ko) [reply nixos/modules/services/misc/transfer-sh.nix:2]: consider `inherits` where you find yourself using `lib.*` excessively.
[this](https://nix.dev/guides/best-practices.html#with-scopes)  nix.dev page on best practices with `with` may be helpful
an issue (#208242) has been made to track inappropriate uses of `with` such as this.

this suggestion has not been resolved.

File: nixos/modules/services/misc/transfer-sh.nix [hunk 28]
```diff
@@ -0,0 +1,422 @@
...
+  options = {
+    services.transfer-sh = {
+      enable = mkEnableOption "transfer-sh setup";
+
+      package = mkPackageOption pkgs "transfer-sh" { };
+
+      environmentFile = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        description = lib.mdDoc "Environment file as defined in {manpage}`systemd.exec(5)`.";
```
Comment (nu-nu-ko) [inline nixos/modules/services/misc/transfer-sh.nix:17]:
```suggestion
        description = "Environment file as defined in {manpage}`systemd.exec(5)`.";
```
`lib.mdDoc` is now just an alias and can be safely removed everywhere.
see https://github.com/NixOS/nixpkgs/commit/d36f950d405369abca5144b14a201bc25c4a080a and https://github.com/NixOS/nixpkgs/pull/237557

File: nixos/modules/services/misc/transfer-sh.nix [hunk 29]
```diff
@@ -0,0 +1,422 @@
...
+      HTTP_AUTH_HTPASSWD = mkOption {
+        description = "htpasswd file path for basic http auth on upload";
+        default = null;
+        type = types.nullOr types.path;
+      };
+
+      HTTP_AUTH_IP_WHITELIST = mkOption {
+        description = "comma separated list of ips allowed to upload without being challenged an http auth";
+        default = [ ];
+        type = with types; listOf str;
```
Comment (nu-nu-ko) [inline nixos/modules/services/misc/transfer-sh.nix:113]: I'd prefer uses of `with` like this are also removed but its much less important than the top most one and is a preference thing.

File: nixos/modules/services/misc/transfer-sh.nix (hunk 22 above)
Comment (nu-nu-ko) [reply nixos/modules/services/misc/transfer-sh.nix:161]: plain text secrets are never fun. If an alternative way to get the secret to the service exists you should certainly be using that, I'd say this should be removed if there is any alternative.

**Review by fsnkty — CHANGES_REQUESTED** [review]

**Comment by pinpox** [comment]
Should I close this PR in favor of https://github.com/NixOS/nixpkgs/pull/283660 ? I don't mind and think it makes sense to decide on one of the two PRs and focus work there to push it over the finish line.

**Pull request closed without merging**